src/mlpack/methods/reinforcement_learning/worker/one_step_q_learning_worker.hpp

```diff
@@ -0,0 +1,172 @@
+/**
+ * @file one_step_q_learning_worker.hpp
+ * @author Shangtong Zhang
+ *
+ * This file is the definition of OneStepQLearningWorker class,
+ * which implements an episode for async one step Q-Learning algorithm.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_RL_WORKER_ONE_STEP_Q_LEARNING_WORKER_HPP
+#define MLPACK_METHODS_RL_WORKER_ONE_STEP_Q_LEARNING_WORKER_HPP
+
+#include <mlpack/methods/reinforcement_learning/training_config.hpp>
+
+namespace mlpack {
+namespace rl {
+
+class OneStepQLearningWorker
+{
+ public:
+  /**
+   * An episode of one step Q-Learning.
+   *
+   * @tparam EnvironmentType The type of the reinforcement learning task.
+   * @tparam NetworkType The type of the network model.
+   * @tparam UpdaterType The type of the optimizer.
+   * @tparam PolicyType The type of the behavior policy.
+   * @param learningNetwork Global learning network.
+   * @param targetNetwork Global target network.
+   * @param stop Whether training is end of not.
+   * @param totalSteps Total time steps of all the workers
+   * @param policy Behavior policy.
+   * @param updater Local optimizer of the worker.
+   * @param environment The reinforcement learning task.
+   * @param config Hyper-parameters for training.
+   * @param deterministic Whether this episode is deterministic or not.
+   * @return Total reward of this episode.
+   */
+  template <
+    typename EnvironmentType,
+    typename NetworkType,
+    typename UpdaterType,
+    typename PolicyType
+  >
+  static double Episode(NetworkType& learningNetwork,
+                        NetworkType& targetNetwork,
+                        bool& stop,
+                        size_t& totalSteps,
+                        PolicyType& policy,
+                        UpdaterType& updater,
+                        EnvironmentType& environment,
+                        const TrainingConfig& config,
+                        bool deterministic) {
+    using StateType = typename EnvironmentType::State;
+    using ActionType = typename EnvironmentType::Action;
+    using TransitionType = std::tuple<StateType, ActionType, double, StateType>;
+
+    updater.Initialize(learningNetwork.Parameters().n_rows,
+        learningNetwork.Parameters().n_cols);
+
+    // Construct the local network from global network.
+    NetworkType network = learningNetwork;
+
+    // Get initial state.
+    StateType state = environment.InitialSample();
+
+    size_t steps = 0;
+    double totalReturn = 0.0;
+
+    // Store pending transitions for update later on.
+    std::vector<TransitionType> pending;
```

Comment:
Do you think, we should reserve the size here, to avoid ``pending.clear();`` in line 148? Depending on how often we call ``clear`` we could speed things up?